### PR TITLE
Bilby data generation

### PR DESCRIPTION
**Pull request by EthanMarx**

### CODE REVIEW

File: projects/sandbox/data_generation/data_generation/bilby_injections.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+    # project raw polarizations onto interferometers
+    # with sampled sky localizations
+    tensors, vertices = get_ifo_geometry(*ifos)
+
+    # dec is declination
+    # psi is polarization angle
+    # phi is relative azimuthal angle between source and earth
+    dec = torch.Tensor(parameters["dec"])
+    psi = torch.Tensor(parameters["psi"])
+    phi = torch.Tensor(parameters["ra"]) - np.pi
```
Comment (deepchatterjeeligo) [inline projects/sandbox/data_generation/data_generation/bilby_injections.py:135]: This is incorrect. What is the coordinate system of get_ifo_geometry

Comment (deepchatterjeeligo) [reply projects/sandbox/data_generation/data_generation/bilby_injections.py:135]: If it is the celestial frame, then the reference for the azimuthal coordinate is gmst. You can convert the time of the injection to gmst either using astropy or lal. I checked that both give consistent value. Ex. the start time in our project config is
```
$ lal_tconvert 1262751694
Sat Jan 11 04:21:16 UTC 2020
```

```
>>> lal.GreenwichMeanSiderealTime(lal.LIGOTimeGPS(1262751694)) % (2*np.pi)
3.0625954275359035
>>> t = Time(1262751694, format='gps', scale='utc')
>>> t.sidereal_time('mean', 'greenwich').to('rad')
<Longitude 3.06258219 rad>
```
which are similar up to 4th decimal place

File: projects/sandbox/data_generation/data_generation/bilby_injections.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+            data, dt=1 / sample_rate, channel=f"{ifo}:{channel}", t0=0
+        )
+
+    data_dict.write(
+        datadir / "bilby_timeseries.hdf5",
+        format="hdf5",
+        overwrite=True,
+    )
+    # save parameters as hdf5 file
+    with h5py.File(datadir / "bilby_injection_parameters.hdf5", "w") as f:
```
Comment (deepchatterjeeligo) [inline projects/sandbox/data_generation/data_generation/bilby_injections.py:175]: Are we saving the data in a format that our sample project also understands?

File: projects/sandbox/data_generation/data_generation/bilby_injections.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+    buffer: float,
+    datadir: Path,
+    logdir: Path,
+    waveform_arguments: Optional[Dict] = None,
+    parameter_conversion: Optional[Callable] = None,
+    gaussian: bool = False,
+    verbose: bool = False,
+    force_generation: bool = False,
+):
+
```
Comment (deepchatterjeeligo) [inline projects/sandbox/data_generation/data_generation/bilby_injections.py:47]: docstring

File: projects/sandbox/data_generation/data_generation/pp_plot_injections.py [hunk 4]
```diff
@@ -177,7 +177,7 @@ def main(
     # phi is relative azimuthal angle between source and earth
     dec = torch.Tensor(parameters["dec"])
     psi = torch.Tensor(parameters["psi"])
-    phi = torch.Tensor(parameters["phi"])
+    phi = torch.Tensor(parameters["ra"]) - np.pi
```
Comment (deepchatterjeeligo) [inline projects/sandbox/data_generation/data_generation/pp_plot_injections.py:180]: See my comment above. The conversions should be consolidated in `utils.py`.

File: projects/sandbox/data_generation/data_generation/pp_plot_injections.py
Comment (deepchatterjeeligo) [inline projects/sandbox/data_generation/data_generation/pp_plot_injections.py:1]: Can we keep the previous name? This does not create a pp-plot.

File: projects/sandbox/data_generation/data_generation/utils.py [hunk 5]
```diff
@@ -93,7 +93,7 @@ def download_data(
     return data.resample(sample_rate)
 
 
-def inject_into_background(
+def inject_into_random_background(
```
Comment (deepchatterjeeligo) [inline projects/sandbox/data_generation/data_generation/utils.py:96]: wondering why the name change here, is this doing anything functionally different now?

File: projects/sandbox/data_generation/data_generation/utils.py [hunk 6]
```diff
@@ -134,3 +134,54 @@ def inject_into_background(
     X += waveforms
 
     return X
+
```
Comment (deepchatterjeeligo) [inline projects/sandbox/data_generation/data_generation/utils.py:137]: Why is this diff showing up as new lines? We already have this function: https://github.com/ML4GW/PE/blob/caa535e73bcf996f4aa7d2d4374c7a017c21b192/projects/sandbox/data_generation/data_generation/utils.py#L96-L136

Comment (deepchatterjeeligo) [reply projects/sandbox/data_generation/data_generation/utils.py:137]: Oh OK. I see there is a `inject_into_background` and a `inject_into_random_background`. What is the difference between them?

File: projects/sandbox/data_generation/tests/test_generate_pp_plot_injections.py [hunk 7]
```diff
@@ -32,39 +30,7 @@ def n_samples(request):
...
-        background,
-        waveforms,
-        kernel_size,
-    )
-
-    assert injections.shape == (n_waveforms, len(ifos), kernel_size)
-
-
-def test_generate_test_set(
+def test_generate_pp_plot_injections(
```
Comment (deepchatterjeeligo) [inline projects/sandbox/data_generation/tests/test_generate_pp_plot_injections.py:33]: Let's use the original naming.

File: projects/sandbox/pyproject.toml [hunk 8]
```diff
@@ -87,35 +93,72 @@ valid_stride = 0.0625
...
+waveform = "mlpe.injection.waveforms.sine_gaussian_frequency"
+n_samples = 8
+waveform_duration = "${base.waveform_duration}"
+spacing = 100
+buffer = 10
+datadir = "${base.datadir}"
+logdir = "${base.logdir}"
+gaussian = false
+
+[tool.typeo.scripts.sample-with-bilby]
```
Comment (deepchatterjeeligo) [inline projects/sandbox/pyproject.toml:132]: this part of the config is for the other merge request. Please remove.

File: projects/sandbox/pyproject.toml [hunk 9]
```diff
@@ -87,35 +93,72 @@ valid_stride = 0.0625
...
+datadir = "${base.datadir}"
+writedir = "${base.datadir}"
+logdir = "${base.logdir}"
+accounting_group = "${LIGO_ACCOUNTING_GROUP}"
+channel = "${base.channel}"
+sample_rate = "${base.sample_rate}"
+prior = "${base.prior}"
 
-[tool.typeo.scripts.sample_testing_set]
+[tool.typeo.scripts.sample-with-flow]
```
Comment (deepchatterjeeligo) [inline projects/sandbox/pyproject.toml:147]: undo these changes as well. They will be done in the next patch for the sampling.

**Review by deepchatterjeeligo — CHANGES_REQUESTED** [review]
@EthanMarx requesting some changes here. Some of them are cosmetic. The primary one related to the content is the conversion between ra and phi in the bilby injections. I have left some example snippets using lal and astropy to get the gmst based on the UTC time. Also, please try to use the bilby conversions as much as possible for consistency: https://git.ligo.org/lscsoft/bilby/-/blob/master/bilby/core/utils/conversion.py#L4

File: projects/sandbox/data_generation/data_generation/bilby_injections.py [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+    state_flag: str,
+    frame_type: str,
+    channel: str,
+    waveform: Callable,
+    spacing: float,
+    start: float,
+    stop: float,
+    sample_rate: float,
+    waveform_duration: float,
+    bilby_duration: float,
```
[the remark this answers is not in the document]
Comment (EthanMarx) [reply projects/sandbox/data_generation/data_generation/bilby_injections.py:37]: The duration of data that bilby analyzes

File: projects/sandbox/data_generation/data_generation/bilby_injections.py [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+        buffer + waveform_duration // 2, n_samples * spacing, n_samples
+    )
+    # bilby requests the start time of the signals
+    # so subtract half the bilby
+    # duration which will enforce that the signal
+    # lies in the center of the kernel
+    np.savetxt(
+        datadir / "start_times.txt", signal_times - (bilby_duration / 2)
+    )
+    min_duration = (
```
[the remark this answers is not in the document]
Comment (EthanMarx) [reply projects/sandbox/data_generation/data_generation/bilby_injections.py:85]: Just to add some space between start (and end) of timeseries and the first (or last) injection

File: projects/sandbox/data_generation/data_generation/bilby_injections.py (hunk 2 above)
Comment (EthanMarx) [reply projects/sandbox/data_generation/data_generation/bilby_injections.py:175]: Yes

File: projects/sandbox/data_generation/data_generation/utils.py (hunk 6 above)
Comment (EthanMarx) [reply projects/sandbox/data_generation/data_generation/utils.py:137]: `inject_into_background` injects waveforms into a timeseries at specific times. This is used for generating injections for bilby.

`inject_into_random_background` randomly samples windows of `kernel_length` from background strain and injects signals into the center of those kernels. This is used for pp-plot injections

File: projects/sandbox/data_generation/data_generation/pp_plot_injections.py
Comment (deepchatterjeeligo) [reply projects/sandbox/data_generation/data_generation/pp_plot_injections.py:1]: Let's use a different name. One that avoids "pp_plot". It's just a denser injection set.

File: projects/sandbox/data_generation/data_generation/utils.py (hunk 6 above)
Comment (EthanMarx) [reply projects/sandbox/data_generation/data_generation/utils.py:138]: They do completely different tasks. `inject_into_random_background` selects many random non-coincident kernels from a long stretch of background. `inject_into_background` injects deterministically into a long stretch of timeseries.

**Review by deepchatterjeeligo — APPROVED** [review]
Looks good.
